samples/extensions/dynamic_multisample_rasterization/dynamic_multisample_rasterization.cpp

```diff
@@ -0,0 +1,947 @@
+/* Copyright (c) 2024, Mobica Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "dynamic_multisample_rasterization.h"
+
+#include "gltf_loader.h"
+#include "scene_graph/components/material.h"
+#include "scene_graph/components/mesh.h"
+#include "scene_graph/components/pbr_material.h"
+#include "scene_graph/components/sub_mesh.h"
+
+DynamicMultisampleRasterization::DynamicMultisampleRasterization()
+{
+	title = "DynamicState3 Multisample Rasterization";
+
+	set_api_version(VK_API_VERSION_1_2);
+	add_instance_extension(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+	add_device_extension(VK_EXT_EXTENDED_DYNAMIC_STATE_3_EXTENSION_NAME);
+	add_device_extension(VK_KHR_DYNAMIC_RENDERING_EXTENSION_NAME);
+}
+
+DynamicMultisampleRasterization::~DynamicMultisampleRasterization()
+{
+	if (device)
+	{
+		vkDestroyPipeline(get_device().get_handle(), pipeline, nullptr);
+		vkDestroyPipeline(get_device().get_handle(), pipeline_inversed_rasterizer, nullptr);
+		vkDestroyPipelineLayout(get_device().get_handle(), pipeline_layout, nullptr);
+		vkDestroyDescriptorSetLayout(get_device().get_handle(), descriptor_set_layout, nullptr);
+
+		vkDestroyPipeline(get_device().get_handle(), pipeline_gui, nullptr);
+		vkDestroyDescriptorSetLayout(get_device().get_handle(), descriptor_set_layout_gui, nullptr);
+		vkDestroyDescriptorPool(get_device().get_handle(), descriptor_pool_gui, VK_NULL_HANDLE);
+		destroy_image_data(depth_stencil);
+		destroy_image_data(color_attachment);
+	}
+}
+
+void DynamicMultisampleRasterization::request_gpu_features(vkb::PhysicalDevice &gpu)
+{
+	// Query the extended dynamic state support
+	VkPhysicalDeviceExtendedDynamicState3FeaturesEXT extended_dynamic_state_3_features{};
+	extended_dynamic_state_3_features.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_EXTENDED_DYNAMIC_STATE_3_FEATURES_EXT;
+
+	VkPhysicalDeviceFeatures2 features2{};
+	features2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FEATURES_2;
+	features2.pNext = &extended_dynamic_state_3_features;
+	vkGetPhysicalDeviceFeatures2(gpu.get_handle(), &features2);
+
+	if (extended_dynamic_state_3_features.extendedDynamicState3RasterizationSamples)
+	{
+		// Only request the features that we support
+		auto &features = gpu.request_extension_features<VkPhysicalDeviceExtendedDynamicState3FeaturesEXT>(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_EXTENDED_DYNAMIC_STATE_3_FEATURES_EXT);
```

Comment:
You're still requesting every supported feature here, as long as extendedDynamicState3RasterizationSamples is supported.
Maybe better:
```
VkPhysicalDeviceExtendedDynamicState3FeaturesEXT requested_feature = {VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_EXTENDED_DYNAMIC_STATE_3_FEATURES_EXT};
requested_feature.extendedDynamicState3RasterizationSamples        = VK_TRUE;
gpu.request_extension_features<VkPhysicalDeviceExtendedDynamicState3FeaturesEXT>(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_EXTENDED_DYNAMIC_STATE_3_FEATURES_EXT) = requested_feature;
```

Reply:
Added simplify version from your proposition.